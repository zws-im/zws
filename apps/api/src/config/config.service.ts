import { Injectable } from '@nestjs/common';
import { url, cleanEnv, json, num, port, str } from 'envalid';

type NodeEnv = 'production' | 'development' | 'staging';

const DEFAULT_SHORT_CHARS: readonly string[] = [
	'\u200C',
	'\u200D',
	'\uDB40\uDC61',
	'\uDB40\uDC62',
	'\uDB40\uDC63',
	'\uDB40\uDC64',
	'\uDB40\uDC65',
	'\uDB40\uDC66',
	'\uDB40\uDC67',
	'\uDB40\uDC68',
	'\uDB40\uDC69',
	'\uDB40\uDC6A',
	'\uDB40\uDC6B',
	'\uDB40\uDC6C',
	'\uDB40\uDC6D',
	'\uDB40\uDC6E',
	'\uDB40\uDC6F',
	'\uDB40\uDC70',
	'\uDB40\uDC71',
	'\uDB40\uDC72',
	'\uDB40\uDC73',
	'\uDB40\uDC74',
	'\uDB40\uDC75',
	'\uDB40\uDC76',
	'\uDB40\uDC77',
	'\uDB40\uDC78',
	'\uDB40\uDC79',
	'\uDB40\uDC7A',
	'\uDB40\uDC7F',
];

export const env = cleanEnv(process.env, {
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	NODE_ENV: str({ default: 'production', choices: ['production', 'development', 'staging'] }),
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	PORT: port({ default: 3000 }),
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	DATABASE_URL: url({ desc: 'PostgreSQL URL' }),
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	SENTRY_DSN: url({ desc: 'Sentry DSN' }),
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	REDIS_URL: url({ desc: 'Redis URL' }),
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	ADMIN_USERNAME: str({ desc: 'Admin username' }),
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	ADMIN_API_TOKEN: str({ desc: 'Admin API token' }),
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	WEBSITE_URL: url({ desc: 'Website URL' }),
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	SHORT_LENGTH: num({ desc: 'Number of characters to generate in a shortened URL', default: 7 }),
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	SHORT_CHARS: json<readonly string[]>({
		desc: 'Characters to use in shortened URLs',
		default: DEFAULT_SHORT_CHARS,
	}),
	// biome-ignore lint/style/useNamingConvention: This is an environment variable
	SHORT_REWRITES: json<Record<string, string>>({
		desc: 'A mapping of characters to apply to short IDs before they are used',
		default: {},
	}),
});

@Injectable()
export class ConfigService {
	public readonly nodeEnv: NodeEnv;
	public readonly port: number;
	public readonly databaseUrl: string;
	public readonly sentryDsn: string;
	public readonly redisUrl: string;
	public readonly adminUsername: string;
	public readonly adminApiToken: string;
	public readonly websiteUrl: string;
	public readonly shortenedLength: number;
	public readonly characters: readonly string[];
	public readonly version = '3.0.0';

	constructor() {
		this.nodeEnv = env.NODE_ENV;
		this.port = env.PORT;
		this.databaseUrl = env.DATABASE_URL;
		this.sentryDsn = env.SENTRY_DSN;
		this.adminUsername = env.ADMIN_USERNAME;
		this.adminApiToken = env.ADMIN_API_TOKEN;
		this.websiteUrl = env.WEBSITE_URL;
		this.redisUrl = env.REDIS_URL;
		this.shortenedLength = env.SHORT_LENGTH;
		this.characters = env.SHORT_CHARS;
	}
}
