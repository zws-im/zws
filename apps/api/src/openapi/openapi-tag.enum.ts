export enum OpenapiTag {
	ShortenedUrls = 'Shortened URLs',
	Badges = 'Badges',
	InstanceStats = 'Instance stats',
	Health = 'Health',
}
