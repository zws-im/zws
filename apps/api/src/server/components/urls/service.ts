import {Buffer} from 'node:buffer';
import {multiReplace, sample} from '@jonahsnider/util';
import type {ShortenedUrl} from '@prisma/client';
import PrismaClientPackage from '@prisma/client';
import * as Sentry from '@sentry/node';
import type {Opaque} from 'type-fest';
import * as config from '../../../config/index.js';

import db from '../../../db.js';
import baseLogger from '../../../logger.js';
import {UniqueShortIdTimeout} from '../../errors.js';

const {ApproximateCountKind} = PrismaClientPackage;

const logger = baseLogger.withTag('services').withTag('urls');
/** Logger for the visits operation. */
const visitLogger = logger.withTag('visit');

/** A short ID. */
export type Short = Opaque<string, 'Short'>;

/** A base64 encoded string. */
type Base64 = Opaque<string, 'Base64'>;

interface Stats {
	url: string;
	visits: Date[];
}

interface VisitUrlData {
	longUrl: string;
	blocked: boolean;
}

/** Maximum number of attempts to generate a unique ID. */
const maxGenerationAttempts = 10;

function encode(value: string): Base64 {
	return Buffer.from(value).toString('base64') as Base64;
}

export function normalizeShortId(id: Short): Short {
	return multiReplace(id, config.characters.rewrites) as Short;
}

/**
 * Generate a short ID.
 * Not guaranteed to be unique.
 *
 * @returns A short ID
 */
function generateShortId(): Short {
	let shortId = '' as Short;

	// eslint-disable-next-line @typescript-eslint/prefer-for-of
	for (let i = 0; i < config.characters.length; i++) {
		// @ts-expect-error String concatenation forces string type
		shortId += sample(config.characters.characters) as Short;
	}

	return shortId;
}

/**
 * Visit a shortened URL.
 *
 * @param id - The ID of the shortened URL to visit
 * @param track - If the visit should be tracked
 *
 * @returns The long URL and if it was blocked
 */
export async function visit(id: Short, track: boolean): Promise<VisitUrlData | null> {
	const encodedId = encode(id);

	const shortenedUrl = await db.shortenedUrl.findUnique({where: {shortBase64: encodedId}, select: {url: true, blocked: true}});

	if (shortenedUrl === null) {
		return null;
	}

	if (track && !shortenedUrl.blocked) {
		// Only track URLs if they aren't blocked
		// This is because we know that a blocked URL won't be sent to clients if they are visiting it

		db.$transaction([
			db.visit.create({data: {shortenedUrl: {connect: {shortBase64: encodedId}}}}),
			db.approximateCounts.update({where: {kind: ApproximateCountKind.VISITS}, data: {count: {increment: 1}}}),
		]).catch(error => {
			Sentry.captureException(error);
			visitLogger.error('failed to create visit', error);
		});
	}

	return {longUrl: shortenedUrl.url, blocked: shortenedUrl.blocked};
}

/**
 * Retrieve usage statistics for a shortened URL
 *
 * @param id - The ID of the shortened URL
 *
 * @returns Shortened URL information and statistics, or `null` if it couldn't be found
 */
export async function stats(id: Short): Promise<null | Stats> {
	const encodedId = encode(id);

	const [visits, shortenedUrl] = await db.$transaction([
		db.visit.findMany({where: {shortenedUrlId: encodedId}, select: {timestamp: true}, orderBy: {timestamp: 'asc'}}),
		db.shortenedUrl.findUnique({where: {shortBase64: encodedId}, select: {url: true}}),
	]);

	if (!shortenedUrl) {
		return null;
	}

	return {visits: visits.map(visit => visit.timestamp), url: shortenedUrl.url};
}

/**
 * Shorten a long URL
 *
 * @param url - The long URL to shorten
 *
 * @returns The ID of the shortened URL
 */
export async function shorten(url: string): Promise<Short> {
	let attempts = 0;
	let created: ShortenedUrl | null = null;
	let id: Short;

	do {
		if (attempts++ > maxGenerationAttempts) {
			throw new UniqueShortIdTimeout(maxGenerationAttempts);
		}

		id = generateShortId();
		const shortBase64 = encode(id);

		try {
			// eslint-disable-next-line no-await-in-loop
			[created] = await db.$transaction([
				db.shortenedUrl.create({data: {url, shortBase64}}),
				db.approximateCounts.update({where: {kind: ApproximateCountKind.SHORTENED_URLS}, data: {count: {increment: 1}}}),
			]);
		} catch {}
	} while (!created);

	return id;
}
