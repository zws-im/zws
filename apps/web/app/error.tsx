'use client';

import { H } from '@highlight-run/next/client';
import { useEffect } from 'react';
import { Button } from './components/button';
import { H1 } from './components/headings/h1';

const SHOULD_REPORT_ERROR = process.env.NODE_ENV !== 'development';

// biome-ignore lint/style/noDefaultExport: This must be a default export
export default function ErrorPage({ error, reset }: { error: Error; reset: () => void }) {
	useEffect(() => {
		if (SHOULD_REPORT_ERROR) {
			H.consumeError(error);
		}
	}, [error]);

	return (
		<main className='flex flex-col items-center justify-center gap-y-4 max-md:pt-48 max-sm:pt-32 md:pt-72'>
			<H1>An unexpected error occurred</H1>
			<Button onClick={reset}>Try again</Button>
		</main>
	);
}
