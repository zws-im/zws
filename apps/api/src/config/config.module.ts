import { Global, Module } from '@nestjs/common';
import { ConfigService } from './config.service';

@Global()
@Module({
	providers: [ConfigService],
	exports: [ConfigService],
})
export class ConfigModule {}
