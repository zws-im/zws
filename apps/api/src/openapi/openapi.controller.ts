import { Controller, Get, Inject, NotFoundException } from '@nestjs/common';
import type { OpenAPIObject } from '@nestjs/swagger';
import { OpenapiService } from './openapi.service';

@Controller('/openapi.json')
export class OpenapiController {
	constructor(@Inject(OpenapiService) private readonly openapiService: OpenapiService) {}

	@Get('/')
	getSpec(): OpenAPIObject {
		const spec = this.openapiService.getSpec();

		if (!spec) {
			throw new NotFoundException('OpenAPI spec not found');
		}

		return spec;
	}
}
