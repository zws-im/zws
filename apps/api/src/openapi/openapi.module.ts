import {Module} from '@nestjs/common';
import {AppConfigModule} from '../app-config/app-config.module';
import {OpenApiService} from './openapi.service';

@Module({
	imports: [AppConfigModule],
	providers: [OpenApiService],
})
export class OpenApiModule {}
