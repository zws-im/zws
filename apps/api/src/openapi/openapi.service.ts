import {Injectable} from '@nestjs/common';
import {DocumentBuilder} from '@nestjs/swagger';
import {paramCase} from 'change-case';
import {AppConfig} from '../app-config/app.config';
import pkg from '../../package.json';

@Injectable()
export class OpenApiService {
	static operationIdFactory(controllerKey: string, methodKey: string) {
		return `${paramCase(controllerKey.replace(/Controller$/, ''))}-${paramCase(methodKey)}`;
	}

	private readonly hostname: string | undefined;
	private readonly port: number;

	constructor(appConfig: AppConfig) {
		this.hostname = appConfig.hostname;
		this.port = appConfig.port;
	}

	getConfig() {
		return new DocumentBuilder()
			.setTitle('Zero Width Shortener')
			.setVersion(pkg.version)
			.setDescription('Shorten URLs with invisible spaces.')
			.setContact(pkg.author.name, pkg.author.url, pkg.author.email)
			.setLicense('Apache 2.0', 'https://www.apache.org/licenses/LICENSE-2.0.html')

			.addBearerAuth({type: 'http'})

			.addServer('http://{host}/{basePath}', 'Custom server (HTTP)', {
				host: {default: this.hostname === undefined ? `localhost:${this.port}` : this.hostname},
				basePath: {default: ''},
			})
			.addServer('https://{host}/{basePath}', 'Custom server (HTTPS)', {
				host: {default: this.hostname === undefined ? `localhost:${this.port}` : this.hostname},
				basePath: {default: ''},
			})
			.addServer('https://api.zws.im', 'zws.im production')

			.addTag('urls', 'Shortened URLs')
			.addTag('stats', 'Usage statistics')
			.addTag('shields', 'Shields endpoint badges')
			.addTag('health', 'Health checks')
			.build();
	}
}
