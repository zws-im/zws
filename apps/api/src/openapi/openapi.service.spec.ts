import type {TestingModule} from '@nestjs/testing';
import {Test} from '@nestjs/testing';
import {AppConfig} from '../app-config/app.config';
import {OpenApiService} from './openapi.service';

describe('OpenApiService', () => {
	let service: OpenApiService;

	beforeEach(async () => {
		const module: TestingModule = await Test.createTestingModule({
			providers: [
				OpenApiService,
				{
					provide: AppConfig,
					useValue: {
						hostname: 'hostname',
						port: 3000,
					},
				},
			],
		}).compile();

		service = module.get<OpenApiService>(OpenApiService);
	});

	describe('getConfig', () => {
		it('generates a config', () => {
			const config = service.getConfig();

			// Version changes very frequently so testing it is very annoying
			config.info.version = 'VERSION';

			expect(config).toMatchSnapshot('config');
		});
	});
});
