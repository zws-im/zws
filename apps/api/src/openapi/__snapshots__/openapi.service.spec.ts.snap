// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OpenApiService getConfig generates a config: config 1`] = `
{
  "components": {
    "securitySchemes": {
      "bearer": {
        "bearerFormat": "JWT",
        "scheme": "bearer",
        "type": "http",
      },
    },
  },
  "info": {
    "contact": {
      "email": "jonah@jonah.pw",
      "name": "Jonah Snider",
      "url": "https://jonahsnider.com",
    },
    "description": "Shorten URLs with invisible spaces.",
    "license": {
      "name": "Apache 2.0",
      "url": "https://www.apache.org/licenses/LICENSE-2.0.html",
    },
    "title": "Zero Width Shortener",
    "version": "VERSION",
  },
  "openapi": "3.0.0",
  "servers": [
    {
      "description": "Custom server (HTTP)",
      "url": "http://{host}/{basePath}",
      "variables": {
        "basePath": {
          "default": "",
        },
        "host": {
          "default": "hostname",
        },
      },
    },
    {
      "description": "Custom server (HTTPS)",
      "url": "https://{host}/{basePath}",
      "variables": {
        "basePath": {
          "default": "",
        },
        "host": {
          "default": "hostname",
        },
      },
    },
    {
      "description": "zws.im production",
      "url": "https://api.zws.im",
      "variables": undefined,
    },
  ],
  "tags": [
    {
      "description": "Shortened URLs",
      "name": "urls",
    },
    {
      "description": "Usage statistics",
      "name": "stats",
    },
    {
      "description": "Shields endpoint badges",
      "name": "shields",
    },
    {
      "description": "Health checks",
      "name": "health",
    },
  ],
}
`;
